marshal_test.go

```diff
@@ -0,0 +1,379 @@
+	method:    "POST",
+	val: func() interface{} {
+		v := struct {
+			F1 string `httprequest:"user,body"`
+		}{
+			F1: "test user",
+		}
+		return &v
+	},
+	expectError: "request wildcards must be named with a non-empty name",
+}, {
+	about:     "nil field to form",
+	urlString: "http://localhost:8081/u",
+	val: func() interface{} {
+		v := struct {
+			F1 *string `httprequest:"user,form"`
+		}{}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/u",
+}, {
+	about:     "nil field to path",
+	urlString: "http://localhost:8081/u",
+	val: func() interface{} {
+		v := struct {
+			F1 *string `httprequest:"user,path"`
+		}{}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/u",
+}, {
+	about:     "marshal to body of a GET request",
+	urlString: "http://localhost:8081/u",
+	val: func() interface{} {
+		v := struct {
+			F1 string `httprequest:"user,body"`
+		}{
+			F1: "hello test",
+		}
+		return &v
+	},
+	expectError: "cannot marshal field: trying to marshal to body of a request with method \"GET\"",
+}, {
+	about:     "marshal to nil value to body",
+	urlString: "http://localhost:8081/u",
+	val: func() interface{} {
+		v := struct {
+			F1 *string `httprequest:"user,body"`
+		}{
+			F1: nil,
+		}
+		return &v
+	},
+	expectBody: "",
+}, {
+	about:     "nil TextMarhsaler",
+	urlString: "http://localhost:8081/u",
+	val: func() interface{} {
+		v := struct {
+			F1 *testMarshaler `httprequest:"surname,form"`
+		}{
+			F1: (*testMarshaler)(nil),
+		}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/u",
+}, {
+	about:     "marshal nil with Sprint",
+	urlString: "http://localhost:8081/u",
+	val: func() interface{} {
+		v := struct {
+			F1 *int `httprequest:"surname,form"`
+		}{
+			F1: (*int)(nil),
+		}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/u",
+}, {
+	about:     "marshal to path with * placeholder",
+	urlString: "http://localhost:8081/u/*name",
+	val: func() interface{} {
+		v := struct {
+			F1 string `httprequest:"name,path"`
+		}{
+			F1: "test",
+		}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/u/test",
+}, {
+	about:     "* placeholder allowed only at the ned",
+	urlString: "http://localhost:8081/u/*name/document",
+	val: func() interface{} {
+		v := struct {
+			F1 string `httprequest:"name,path"`
+		}{
+			F1: "test",
+		}
+		return &v
+	},
+	expectError: `placeholders starting with \* are only allowed at the end`,
+},
+}
+
+func (*marshalSuite) TestMarshal(c *gc.C) {
+	for i, test := range marshalTests {
+		c.Logf("%d: %s", i, test.about)
+		method := "GET"
+		if test.method != "" {
+			method = test.method
+		}
+		req, err := httprequest.Marshal(test.urlString, method, test.val())
+		if test.expectError != "" {
+			c.Assert(err, gc.ErrorMatches, test.expectError)
+			continue
+		}
+		c.Assert(err, gc.IsNil)
+		if test.expectURLString != "" {
+			c.Assert(req.URL.String(), gc.DeepEquals, test.expectURLString)
+		}
+		//if test.expectBody != "" {
```

Comment:
if there's an expected body, we should check that the Content-Type header is set to application/json here.